File: Strategy/StrategyPattern.py
from abc import ABCMeta, abstractmethod


class Strategy(object):
    __metaclass__ = ABCMeta

    @abstractmethod
    def execute_algorithm(self):
        pass


class ConcreteStrategy1(Strategy):
    def execute_algorithm(self):
        print("ConcreteStrategy1")


class ConcreteStrategy2(Strategy):
    def execute_algorithm(self):
        print("ConcreteStrategy2")


class Context(object):
    def __init__(self, strategy):
        self._strategy = strategy

    def set_strategy(self, strategy):
        self._strategy = strategy

    def execute(self):
        self._strategy.execute_algorithm()


if __name__ == "__main__":
    strategy1 = ConcreteStrategy1()
    strategy2 = ConcreteStrategy2()

    context = Context(strategy1)
    context.execute()

    context.set_strategy(strategy2)
    context.execute()

File: Strategy/StrategyPattern.java
interface Strategy {
    public void executeAlgorithm();
}

class ConcreteStrategy1 implements Strategy {
    public void executeAlgorithm() {
        System.out.println("ConcreteStrategy1");
    }
}

class ConcreteStrategy2 implements Strategy {
    public void executeAlgorithm() {
        System.out.println("ConcreteStrategy2");
    }
}

class Context {
    private Strategy strategy;

    public Context(Strategy strategy) { 
        this.strategy = strategy;
    }

    public void setStrategy(Strategy strategy) {
        this.strategy = strategy;
    }

    public void execute() {
        strategy.executeAlgorithm();
    }
}

public class StrategyPattern {
    public static void main(String[] args) {
        Strategy strategy1 = new ConcreteStrategy1();
        Strategy strategy2 = new ConcreteStrategy2();

        Context context = new Context(strategy1);
        context.execute();

        context.setStrategy(strategy2);
        context.execute();
    }
}

